src/sorting/python/merge_sort/src/merge_sort.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 17 01 24
+
+@author: tvsirius
+"""
+
+
+def merge_sort(list_to_sort: list[any], ascending: bool = True) -> list[any]:
+    """Sorts a list using merge algorithm
+    Can sort any list of items that can be compared
+    (int, bool,char,**). Items compatibility will be checked on the fly,
+    for performance reasons
+
+    Parameters:
+        list_to_sort: list[any] - List (or tuple) of similar item 
+            that allows comparison methods (< > <= >=)
+        ascending:bool=True - If True (by default) list will be sorted in ascending order, 
+            otherwise descending
+
+    Returns -> list[any] sorted list 
+
+    >>> merge_sort([])
+    []
+
+    >>> merge_sort([3,2])
+    [2,3]
+
+    >>> merge_sort([1,5,10,2,22],ascending = False)
+    [22,10,5,2,1]
+    
+    
+    """
+
+    if len(list_to_sort) == 0:
+        return []
+
+    assert isinstance(list_to_sort, (list, tuple)
+                      ), "list_to_sort must be list or tuple"
+    # I will not test all items compatibility,
+    # I will assert first element for allowing comparison methods (< > <= >=)
+
+    if not (hasattr(list_to_sort[0], '__lt__') and
+            hasattr(list_to_sort[0], '__le__') and
```

Comment:
Even though I can see what this snippet does, I am confused about how it works. I used to think that we used these methods in defining classes.

Reply:
It checking attributes, but it always will be false. 
@tvsirius, are you sure you need this check? Can you give example of type without such attributes? I seems cant find any

Reply:
Yes, you cannot compare strings.
There can be custom classes with implementations of ><>=<=, and can be without. So I want sort to work with those, who can

Other thinf this is a check for first elements, and others are checked with try-except later, so I don't know if it is a good way, but it works

Reply:
Check it out:

print("bcd">"abc")
print("abc">"bcd")

As for custom classes. Every clas is a child of Object class, and object class have all these methods. Custom class maybe not implemented these methods but already has them by default. 

Looks like all types in Python is classes. Even primitives as int, bool, float. I have no idea where function hasattr can be used. Looks like it redundant, and exist fo some comparability with old code. 

Can you rework those checks? You can take idea of how I implemented it in my pull request

Reply:
But you implemnted it wrong, because you cannot sort (int, float) list

I think i can remove this first check, because I have do it with 
try - except block inside code

Reply:
I realy don't know, maybe we all have the point. 
Just built in .sort() and sorted() functions in python, as i know do sort of everything,
and even your code will do it, if it goes over assertions

I have the same issue with PS, as I remember.  I have used  round, multiplication by 1000 and rounding again, just to pass the grader

ok, then, we can ask @colevandersWands about this, and leave our solutions as they are

Reply:
You cannot compare dicst for example
>>> {"f":"f"}<{} 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'dict' and 'dict'
>>>

Reply:
You not supposed to compare two dic at all or lists or tuples. It doesn't have any sense.

You may compare entire collections, but you need to do some math with it's elements before

Reply:
Yes, of course. 
But it's the defensive programming thing we are talking about

Reply:
I would say this is a case where you can decide how you want it to work.  The original tests didn't specify so you get to decide.  If you want a function that only compares the same type, you can have a defensive check to make sure they're the same type (and a matching unit test).  If you want it to only work for eg. ints you can do a defensive type check (and a matching unit test).

What ever you do, just agree to it in code review and document this behavior with type signatures, docstrings and unit tests.

Reply:
Yes, that's what I also thinks. I think for this challenge it's ok to go different ways, just we should document it. So I will update README.md for this challenge.
For the real project we should be more specific, but there we will have the context, that will set our constrains

Reply:
Oh, it's already documented

Reply:
> I would say this is a case where you can decide how you want it to work. The original tests didn't specify so you get to decide. If you want a function that only compares the same type, you can have a defensive check to make sure they're the same type (and a matching unit test). If you want it to only work for eg. ints you can do a defensive type check (and a matching unit test).
> 
> What ever you do, just agree to it in code review and document this behavior with type signatures, docstrings and unit tests.
@colevandersWands 
Yes. But point is that piece of code will be never reached because all data in Python has that attributes. They might be not implemented but hasattr(arr[i]) will always be true and making it inside if not block will make it false which means this block will never be entered by interpreter.

Reply:
There are data types without '__lt__' '__le__'... attributes

Reply:
> There are data types without '**lt**' '**le**'... attributes

Do you have examples?

Reply:
Check this out

isinstance(1, object) # True
isinstance('Hello World', object) # True
isinstance(int, object) # True
isinstance(object, object) # True
isinstance(type, object) # True

https://stackoverflow.com/a/49143578

Every type in Python is a subclass of Object. Object have all these attributes, and children classes as well

Reply:
@Vlad421  good point, I wasn't thinking of it at that level.  In that case it is extra code making the solution harder to understand

Reply:
What I propose is replace attributes test for type test. To exclude mixed type inside data collection, and exclude everything except tuples and lists. 
We could sort dictionaries but for dictionaries we need one more parameter for function to specify sorting by keys or values. Which is redundant it current scope. Even in Python for sorting dict you create 2 lists from keys and values, sort them simultaneously and then combine to dict

Reply:
> Check this out
> 
> isinstance(1, object) # True isinstance('Hello World', object) # True isinstance(int, object) # True isinstance(object, object) # True isinstance(type, object) # True
> 
> https://stackoverflow.com/a/49143578
> 
> Every type in Python is a subclass of Object. Object have all these attributes, and children classes as well

First it was about this specific attiributes:
    if not (hasattr(list_to_sort[0], '__lt__') and
            hasattr(list_to_sort[0], '__le__') and ...
and you are showing that everything is an isstance of object and it's not related here

There is onbject that cannot be compared

As we said before it's a matter of convention, for here we don't have strict rules. I pointed out in README my sort can sort everything that can be compared, in your sort you even prohibit sorting mixed list of int and float
I think we are both fine

I also removed the hasattr(list_to_sort[0], '__lt__') beacuse I am checking it directly with try except block


when we will write function for DS project we will have context that will dictate how we should make any functions

Reply:
>There is onbject that cannot be compared 
>(i.e. dict, and many custom classes)

Dict is collection. It can't be compared. You have to understand it.
Custom classes. Every class is subclass of Object class. Even if you didn't tell it explicitly

Reply:
List and tuples is collections too. We comparing elements inside collection not collections between eas other

Reply:
@tvsirius 
https://docs.python.org/3/reference/datamodel.html#basic-customization

Docs for Object class

Reply:
>>> Custom classes. Every class is subclass of Object class. Even if you didn't tell it explicitly
To this I agree

>>> We are comparing elements inside collection not collections between each other

Yes, so the point was what data types cannot be compared.

our input list to sort should be list or tuple
elements of this list must allow comapison. if we have a list of int,chars,floats,string even list of lists,sets of tuple and any custom type that allows comparision it will sort it

if elemets of list cannot be compared, such as if we pass a list of dicts, or some user type it will return a TypeErorr

Reply:
Also some code snippet for not believers 😁


class Test:
    def __init__(self):
        print("Hello")
        
        
a = Test()

print(isinstance(a, object))
print(isinstance(5, object ))
print(isinstance('a', object))

Man.. Python is fully OOP language. With unfortunately not best OOP implementation 😂

Reply:
> Also some code snippet for not believers 😁

It seems you got me wrong somehow, i was not aguing that everything is an object
read my previous comment

Reply:
Code in if block is unreachable. It will never enters inside that block. This is a point. You need to think about it. It actually great opportunity to learn something new. 

Look at you code and try to play it inside you head. Step by step. You will see

Reply:
To simplify it look like this:
if not true:  #false all time 
    Exception

Reply:
yes, it seems you right, maybe every object has this attributes, but for non comparatable they return an errors
```def test_comparison(obj_test):
    return (hasattr(obj_test, '__lt__') and
            hasattr(obj_test, '__le__') and
            hasattr(obj_test, '__gt__') and
            hasattr(obj_test, '__ge__'))


print(test_comparison('1'))
print(test_comparison(3))
print(test_comparison(test_comparison))
print(test_comparison(print))
print(test_comparison(object))
print(test_comparison({"a":"1","v":3}))
print(test_comparison('2f'))
print(test_comparison([1, 2,4]))
```
True all the time

Reply:
but this code has been already removed

Reply:
> True all the time
Exactly. True all the time. And you explicit 
If **not**. Pay attention you have not statement which reverse Boolean value to opposite. Turns it to False. Which means it will never enter that block to throw exception. And made no use for that check. Because it will never happen. This is what I tell about

Reply:
Ok, now i got what you mean

Reply:
Dunno why you vere provings everything is an object )

Reply:
> Dunno why you vere provings everything is an object )

Because it how Python is made. Funny language.

Reply:
Yeah
I like it a lot